Dolphin scripts/Entrance Randomizer/lib/entrance_rando.py

```diff
@@ -55,79 +91,198 @@ def highjack_transition_rando():
+def get_random_redirection(original: Transition, all_redirections: Iterable[Transition]):
+    possible_redirections = [
+        redirect for redirect in all_redirections
+        if original.from_ != redirect.to  # Prevent looping on itself
+    ]
+    if len(possible_redirections) > 0:
+        return random.choice(possible_redirections)
+    return None
 
 
 def set_transitions_map():
-    _possible_transitions_bucket = list(starmap(Transition, ALL_POSSIBLE_TRANSITIONS))
-    """A temporary container of transitions to pick from until it is empty."""
     transitions_map.clear()
-    for area in TRANSITION_INFOS_DICT.values():
-        for to_og in (exit_.area_id for exit_ in area.exits):
-            original = Transition(from_=area.area_id, to=to_og)
-            redirect = get_random_redirection(original, _possible_transitions_bucket)
-            if redirect is None or original in transitions_map:
-                # Don't override something set in a previous iteration,
-                # like from linked two-way entrances.
+    starting_default = TRANSITION_INFOS_DICT[starting_area].default_entrance
+    tutorial_original = Transition(from_=LevelCRC.JAGUAR, to=LevelCRC.PLANE_CUTSCENE)
+    tutorial_redirect = Transition(from_=starting_default, to=starting_area)
+    transitions_map[tutorial_original] = tutorial_redirect
+
+    _possible_redirections_bucket = list(starmap(Transition, ALL_POSSIBLE_TRANSITIONS))
+    one_way_list = [
+        Transition(from_=LevelCRC.WHITE_VALLEY, to=LevelCRC.MOUNTAIN_SLED_RUN),
+        Transition(from_=LevelCRC.MOUNTAIN_SLED_RUN, to=LevelCRC.APU_ILLAPU_SHRINE),
+        Transition(from_=LevelCRC.APU_ILLAPU_SHRINE, to=LevelCRC.WHITE_VALLEY),
+        Transition(from_=LevelCRC.CAVERN_LAKE, to=LevelCRC.JUNGLE_CANYON),
+    ]
+    if CONFIGS.LINKED_TRANSITIONS:
+        # Ground rules:
+        # 1. you can't make a transition from a level to itself
+        # 2. any 2 levels may have a maximum of 1 connection between them (as long as it's 2-way)
+
+        _possible_origins_bucket = list(starmap(Transition, ALL_POSSIBLE_TRANSITIONS))
+
+        level_list = []
+        for area in TRANSITION_INFOS_DICT.values():
+            area.con_left = len(area.exits)
+            if area.con_left > 0:
+                level_list.append(area)
+        random.shuffle(level_list)
+        level_list.sort(key=lambda a: a.con_left, reverse=True)
+
+        link_list = [link_two_levels(level_list[0], level_list[1])]
+        global total_con_left
+        total_con_left = level_list[0].con_left + level_list[1].con_left
+
+        index = 2
+        while index < len(level_list):
+            r = random.choice(["1", "2"])
+            if total_con_left > 0 and (level_list[index].con_left == 1 or r == "1"):
+                # option 1: connect to one or more existing levels
+                connect_to_existing(level_list, index, link_list)
+            elif level_list[index].con_left > 1 and (total_con_left == 0 or r == "2"):
```

Comment:
I started with your suggestion A, but then the linting programs started complaining that I was comparing "magic numbers", so that's the entire reason I changed to strings in the first place.
I'm not a fan of your suggestion B, that really makes it seem there is some math being done here, while there is only a choice that needs to be made.

I haven't changed it for now.

Reply:
There's always math being done in randomness ;)

For readability you could add a `random_bool` utility function, or `bool(random.randint(0, 1))` that does the same thing (just slightly less efficiently, which doesn't really matter)

>  the linting programs started complaining that I was comparing "magic numbers"

I mean, it's not wrong. Now you're still comparing "magic strings" whose values still don't mean much :P
You can just `# noqa: PLR2004` the line with `== 2`. Or the "correct fix" would be to extract `1` and `2` to meaningful variable names (for example: `CHOICE_INSERT_BETWEEN`, `CHOICE_CONNECT` or `CHOICE_NOT_INSERT_BETWEEN`

(if you wanna be pedantic about it, "choice" should be its own subtype of string, could be an IntEnum as that has finite values)

But if by principle this value should never be anything else than 2 choices, than a boolean is even better.

Reply:
Added a 'Choice' enum that takes care of it for upcoming commit.